### PR TITLE
Add top-level jsonapi member

### PR DESCRIPTION
**Pull request by bf4**
Ref: http://jsonapi.org/format/#document-jsonapi-object

- `jsonapi_include_toplevel_object`: Whether to include a [top level JSON API member](http://jsonapi.org/format/#document-jsonapi-object)
   in the response document.
   Default: `false`.
- Used when `jsonapi_include_toplevel_object` is `true`:
  - `jsonapi_version`: The latest version of the spec the API conforms to.
    Default: `'1.0'`.
  - `jsonapi_toplevel_meta`: Optional metadata. Not included if empty.
    Default: `{}`.

Example meta usage:

```ruby
       
              jsonapi_include_toplevel_object = true,
              jsonapi_toplevel_meta = {
                'copyright' => 'Copyright 2015 Example Corp.'
              }

```

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Yup! I'm interest on this @bf4

**Comment by bf4** [comment]
This is a little harder than I thought due the somewhat recursive way adapters work.

**Comment by joaomdmoura** [comment]
No worries, take your time, let me know if there is something I can do in order to help you.

**Comment by bf4** [comment]
Well, I think I'm going to work on other stuff for a bit.  I was just looking at warnings, and also at how using `class Adapter` as both a namespace and subclass causes some circular requires, and if maybe we should separate `AbstractAdapter` out of `Adapter`, leaving just the `create` and `adapter_class` methods (put those added by the  https://github.com/rails-api/active_model_serializers/pull/1017 )

**Comment by joaomdmoura** [comment]
Yep, please clean our warnings would be great! :smile:

File: lib/active_model/serializer/adapter/json_api.rb [hunk 1]
```diff
@@ -54,7 +78,7 @@ def serializable_hash_for_single_resource(serializer, options)
           primary_data = primary_data_for(serializer, options)
           relationships = relationships_for(serializer)
           included = included_for(serializer)
-          hash = { data: primary_data }
+          hash[:data] = primary_data # FIXME: should be an [], not a {}
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: ugh

File: docs/general/configuration_options.md [hunk 2]
```diff
@@ -9,3 +9,9 @@ The following configuration options can be set on `ActiveModel::Serializer.confi
 ## JSON API
 
 - `jsonapi_resource_type`: Whether the `type` attributes of resources should be singular or plural. Possible values: `:singular, :plural`. Default: `:plural`.
+- `jsonapi_toplevel_member`: Whether to include a [top level JSON API member](http://jsonapi.org/format/#document-jsonapi-object)
```
Comment (bf4) [inline docs/general/configuration_options.md]: defined in JsonApi adapter

File: test/adapter/json_api/toplevel_jsonapi_test.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            end
+          end
+
+          def test_toplevel_jsonapi_meta
+            new_config = {
+              jsonapi_toplevel_member: true,
+              jsonapi_toplevel_meta: {
+                'copyright': 'Copyright 2015 Example Corp.'
+              }
+            }
```
Comment (bf4) [inline test/adapter/json_api/toplevel_jsonapi_test.rb]: The value of meta must be a hash

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,4 +11,5 @@
   * remove root key option and split JSON adapter [@joaomdmoura]
   * adds FlattenJSON as default adapter [@joaomdmoura]
   * adds support for `pagination links` at top level of JsonApi adapter [@bacarini]
-  * adds extended format for `include` option to JSONAPI adapter [@beauby]
+  * adds extended format for `include` option to JsonApi adapter [@beauby]
+  * adds support for top level jsonapi member support [@beauby]
```
Comment (beauby) [inline CHANGELOG.md]: There's an extraneous "support" at the end. Plus it should read `[@beauby, @bf4]` :)

File: lib/active_model/serializer/adapter/json_api.rb (hunk 1 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: n/m, I guess this is legal

> Primary data MUST be either:

> - a single resource object, a single resource identifier object, or null, for requests that target single resources
> - an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: It is, indeed.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: More than legal, it is actually the right behavior when targeting a single resource.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: learning..

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: rebased to remove comment, but this comment remains...

File: lib/active_model/serializer/adapter/json_api.rb [hunk 5]
```diff
@@ -1,10 +1,36 @@
...
+          ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
+          module JsonApi
+            module_function
+
+            def add!(document)
+              return document unless ActiveModel::Serializer.config.jsonapi_toplevel_member
+              object = {
+                jsonapi: {
+                  version: ActiveModel::Serializer.config.jsonapi_version,
+                  meta: ActiveModel::Serializer.config.jsonapi_toplevel_meta
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Meta is optional as per the spec, so it should't be included by default.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 6]
```diff
@@ -1,10 +1,36 @@
...
+                }
+              }
+              object[:jsonapi].reject! { |_, v| v.blank? }
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Aaaand I hadn't read that line yet.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 7]
```diff
@@ -1,10 +1,36 @@
...
+            end
+          end
+        end
         extend ActiveSupport::Autoload
         autoload :PaginationLinks
         autoload :FragmentCache
 
         def initialize(serializer, options = {})
           super
+          @hash = {}
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the instance variable? I think it adds confusion.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 8]
```diff
@@ -1,10 +1,36 @@
 class ActiveModel::Serializer::Adapter::JsonApi < ActiveModel::Serializer::Adapter
+        # JsonApiObject: namespace for functions that build
+        # JSON API objects
+        module JsonApiObject
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: This could be extracted to `adapter/json_api/json_api_object.rb`, what do you think?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 9]
```diff
@@ -1,10 +1,36 @@
...
+          # Make JSON API top-level jsonapi member opt-in
+          # ref: http://jsonapi.org/format/#document-top-level
+          ActiveModel::Serializer.config.jsonapi_toplevel_member = false
+          ActiveModel::Serializer.config.jsonapi_version = '1.0'
+          ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
+          module JsonApi
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the `JsonApi` namespace?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 10]
```diff
@@ -1,10 +1,36 @@
...
+          end
+        end
         extend ActiveSupport::Autoload
         autoload :PaginationLinks
         autoload :FragmentCache
 
         def initialize(serializer, options = {})
           super
+          @hash = {}
+          JsonApiObject::JsonApi.add!(hash)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I'd much rather have it as a conditional assignation than a magical `add!` call that may or may not add stuff to its argument. Moreover, I think it belongs to the `serializable_hash` method rather than the constructor. How about:
```ruby
hash[:jsonapi] = JsonApiObject.serializable_hash if ActiveModel::Serializer.config.jsonapi_toplevel_member
```

Plus I'm starting to wonder whether the config should be named `jsonapi_toplevel_member` or something more explicit like `include_jsonapi_object`.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 11]
```diff
@@ -51,10 +77,11 @@ def serializable_hash_for_collection(serializer, options)
         end
 
         def serializable_hash_for_single_resource(serializer, options)
+          hash[:data] = {}
           primary_data = primary_data_for(serializer, options)
           relationships = relationships_for(serializer)
           included = included_for(serializer)
-          hash = { data: primary_data }
+          hash[:data].merge!(primary_data)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the `merge`?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 8 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Yeah, just not sure if we'll go with this as a pattern. So, I'm incubating it here

File: lib/active_model/serializer/adapter/json_api.rb (hunk 9 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: It's the `JsonApi` Object in the JSON API

File: lib/active_model/serializer/adapter/json_api.rb (hunk 10 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: good point

File: lib/active_model/serializer/adapter/json_api.rb (hunk 11 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Why not?

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Because it is more confusing than an assignation.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 9 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Yeah, but it's already the `JsonApiObject` Object in AMS, so why double namespace?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 11 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: But it's not an assignation... it's a mutation

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: C.f. my other comment, I think it should not be. I don't think there's any valid reason to start building the hash in the constructor as it is the sole responsibility of the `serializable_hash` method.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: On a technical note, it's a mutation solely because you assign it to `{}` right before. Maybe a bit of context will be helpful: the main goal of the JsonApi refactor I did was to decouple methods as much as possible so that they don't modify their arguments and just return the value they're supposed to compute. Hence, there is no need to either initialize `hash[:data]` or `merge` anything into it.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 12]
```diff
@@ -51,10 +77,11 @@ def serializable_hash_for_collection(serializer, options)
         end
 
         def serializable_hash_for_single_resource(serializer, options)
+          hash[:data] = {}
           primary_data = primary_data_for(serializer, options)
           relationships = relationships_for(serializer)
           included = included_for(serializer)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Those three methods have no side-effect on `@hash`.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 9 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: good call. my intention was that I was considering having a whole bunch of `JsonApiObjects` like `JsonApi`, `Link`, `Error`, `Meta`, `Data`, etc.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 13]
```diff
@@ -31,14 +48,26 @@ def fragment_cache(cached_hash, non_cached_hash)
           ActiveModel::Serializer::Adapter::JsonApi::FragmentCache.new.fragment_cache(root, cached_hash, non_cached_hash)
         end
 
+        def fieldset
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: This probably belongs to a different PR.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 14]
```diff
@@ -6,19 +6,36 @@ class JsonApi < Adapter
...
+              }
+              object[:jsonapi].reject! { |_, v| v.blank? }
+              document.merge!(object)
+            end
           end
         end
 
         def serializable_hash(options = nil)
           options ||= {}
+          self.hash = {}
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the instance variable/side-effects?

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Do you mean why `self.hash = ` vs. `@hash =`?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 13 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Yeah, I suppose

File: lib/active_model/serializer/adapter/json_api.rb (hunk 14 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: No I mean, why work on a `self.hash`/`@hash` instead of building an object in the method and returning it?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 15]
```diff
@@ -123,10 +154,13 @@ def relationship_value_for(serializer, options = {})
...
 
         def relationships_for(serializer)
-          Hash[serializer.associations.map { |association| [association.key, { data: relationship_value_for(association.serializer, association.options) }] }]
+          Hash[serializer.associations.map { |association|
+            [association.key, { data: relationship_value_for(association.serializer, association.options) }]
+          }]
         end
 
         def included_for(serializer)
+          included = ActiveModel::Serializer::Utils.include_args_to_hash(instance_options[:include])
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: This will recompute the hash each time `serializable_hash` gets called. Granted, currently it should only happen once, but there might be outside-Rails use-cases where it's not the case.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 16]
```diff
@@ -6,19 +6,36 @@ class JsonApi < Adapter
...
+              object[:jsonapi].reject! { |_, v| v.blank? }
+              document.merge!(object)
+            end
           end
         end
 
         def serializable_hash(options = nil)
           options ||= {}
+          self.hash = {}
+          ApiObjects::JsonApi.include_member? && ApiObjects::JsonApi.add!(hash)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I'd rather have it like that:
```ruby
hash.merge!(ApiObjects::JsonApi.member) if ApiObjects::JsonApi.include_member?
```

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I like reading left to right, for the most part, and rubocop actually recommended this

File: docs/general/configuration_options.md [hunk 17]
```diff
@@ -9,3 +9,11 @@ The following configuration options can be set on `ActiveModel::Serializer.confi
 ## JSON API
 
 - `jsonapi_resource_type`: Whether the `type` attributes of resources should be singular or plural. Possible values: `:singular, :plural`. Default: `:plural`.
+- `jsonapi_include_toplevel_member`: Whether to include a [top level JSON API member](http://jsonapi.org/format/#document-jsonapi-object)
+   in the response document.
+   Default: `false`.
+- Used when `jsonapi_include_toplevel_member` is `true`:
+  - `jsonapi_version`: The latest version of the spec the API conforms to.
+    Default: `'1.0'`.
+  - `jsonapi_toplevel_meta`: Optional metadata. Not included if empty.
```
Comment (bf4) [inline docs/general/configuration_options.md]: I do think there's a latent `json_api` config object this exposes

File: lib/active_model/serializer/adapter/json_api.rb [hunk 18]
```diff
@@ -6,26 +6,56 @@ class JsonApi < Base
...
         autoload :FragmentCache
 
-        def initialize(serializer, options = {})
+        module ApiObjects
+          # Make JSON API top-level jsonapi member opt-in
+          # ref: http://jsonapi.org/format/#document-top-level
+          ActiveModel::Serializer.config.jsonapi_include_toplevel_member = false
+          ActiveModel::Serializer.config.jsonapi_version = '1.0'
+          ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
+          module JsonApi
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Probably is a latent jsonapi config object here, or at least `JsonApi.config.version` via `class JsonApi < Base; def config; ActiveModel::Serializer.config.jsonapi; end;`

File: lib/active_model/serializer/adapter/json_api.rb (hunk 16 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I don't mind it being `ApiObjects::JsonApi.include_member? && hash.merge!(ApiObjects::JsonApi.member)`, I just dislike the unnecessary side-effect.

File: docs/general/configuration_options.md (hunk 17 above)
Comment (beauby) [inline docs/general/configuration_options.md]: Indeed there is.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 19]
```diff
@@ -6,6 +6,32 @@ class JsonApi < Base
...
+            module_function
+
+            def include_member?
+              ActiveModel::Serializer.config.jsonapi_include_toplevel_member
+            end
+
+            def add!(document)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the side-effects :crying_cat_face:

File: lib/active_model/serializer/adapter/json_api.rb [hunk 20]
```diff
@@ -21,11 +47,17 @@ def initialize(serializer, options = {})
         def serializable_hash(options = nil)
           options ||= {}
 
-          if serializer.respond_to?(:each)
-            serializable_hash_for_collection(options)
-          else
-            serializable_hash_for_single_resource(options)
-          end
+          hash =
+            case
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Why the extra line for a `case` statement? I generally like it, but this really looks like an `if...else` situation to me.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: what side-effects?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 20 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: When I was a young dev I learned to dislike using the return value from and if/else, since it's not as comfortable as using the return value from a case/switch

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Since if/else is usually used for control flow

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: alternatively, 

```ruby
if ...
  hash = ...
else 
  hash = ...
end
```

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Having `add!` modify its argument instead of returning a value.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 20 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I think the StyleGuide advises against the latter, and actually encourages using the return value of `if/else`, which is more than common in functional languages.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: But, isn't this question the same as why do I prefer `foo.merge!(bar)` over `foo.merge!(build_bar)`?

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Hmm not really, to me, it's more a matter of `do_some_magic(foo)` vs `foo[:attribute] = compute_something`.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: With the former, in case of bug, you have to check how exactly is `do_some_magic` modifying its argument, whereas with the latter, you can directly see whether it potentially has anything to do with it at all.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: So, what do you want?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 20 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I don't really care what the style guide says, honestly, if we agree on something that makes sense to us.  It does mention 'leveraging' the return value of if/else and case statements.

Is this a blocker?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I think `do_some_magic(foo)` is disingenous since the method is `JsonApiObject.add!(foo)`, it's clear you're mutating foo to addd a JsonApiObject.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 20 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I feel quite strongly about it. I personally think we should leverage the functional constructs of ruby in order to have the clearest/"no surprise-est" code possible, and I think using the return value of an `if/else` allows for that.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: True, but it is not clear what "adding" means, it could be setting `foo[:data][:jsonapi]` for instance.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Or, its job is to add and you don't need to worry about it ;)

File: lib/active_model/serializer/adapter/json_api.rb (hunk 18 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Agreed, as mentioned in https://github.com/rails-api/active_model_serializers/issues/1097, we should somehow namespace the config options.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 19 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Would you mind just adding a commit to my branch?  You obviously have a pretty strong opinions I don't share, so it would be faster if you changed it

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I really don't care that strongly since I see it as internal for now till we figure out how we'd like to use it, if at all

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Will do immediately.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Done.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Since you have the (first and) last commit, do you might (rebasing against master and) getting it passing on ci?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 21]
```diff
@@ -21,11 +46,16 @@ def initialize(serializer, options = {})
...
-            serializable_hash_for_single_resource(options)
-          end
+          hash =
+            if serializer.respond_to?(:each)
+              serializable_hash_for_collection(options)
+            else
+              serializable_hash_for_single_resource(options)
+            end
+
+          ApiObjects::JsonApi.include_member? && hash[:jsonapi] = ApiObjects::JsonApi.member
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: So, the one thing I was trying to avoid is `hash[:jsonapi] =` give the Adapter#serializable_hash the responsibility for knowing the jsonapi object key.  I was hoping the JsonApiObject would maintain all knowledge of building that object

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: that is, I was thinking that each ApiObject would know how to build the json api object it represents

File: lib/active_model/serializer/adapter/json_api.rb [hunk 22]
```diff
@@ -21,11 +48,16 @@ def initialize(serializer, options = {})
...
-            serializable_hash_for_single_resource(options)
-          end
+          hash =
+            if serializer.respond_to?(:each)
+              serializable_hash_for_collection(options)
+            else
+              serializable_hash_for_single_resource(options)
+            end
+
+          ApiObjects::JsonApi.include_member? && hash.merge!(ApiObjects::JsonApi.member)
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Do you prefer it like this?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 18 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Maybe we should do that in this PR before its too late?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 22 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I do, esp. if you're ok with it... I'm willing to be persuaded... the whole ApiObjects::JsonApi is just an idea I'm playing with.. not very useful if it's the only one, but might lead to good abstractions for other objects, like errors, paginations, meta, links, etc.

What do you think of renaming `member` to `object`?  It would align with serializer.object and make sense in context.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 18 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Yeah, it would make sense. Any specific way you'd like it to be?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 22 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I'm ok with renaming `member` to `object`, although both those names are a bit confusing.

File: lib/active_model/serializer/adapter/json_api/api_objects.rb [hunk 23]
```diff
@@ -0,0 +1,38 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module ApiObjects
+          # Make JSON API top-level jsonapi member opt-in
+          # ref: http://jsonapi.org/format/#document-top-level
+          ActiveModel::Serializer.config.jsonapi_include_toplevel_object = false
+          ActiveModel::Serializer.config.jsonapi_version = '1.0'
+          ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: I could see a race condition here now that it's lazy loaded.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: (btw, the reason I didn't originally create this file was because I wasn't sure if I liked the abstraction of `ApiObjects` and wanted it to incubate a bit)

File: lib/active_model/serializer/adapter/json_api/api_objects.rb [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+          module JsonApi
+            module_function
+
+            def add!(hash)
+              hash.merge!(object) if include_object?
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: IMHO, this should either return the resulting hash, or nil, but not both

File: lib/active_model/serializer/configuration.rb [hunk 25]
```diff
@@ -4,6 +4,8 @@ module Configuration
       include ActiveSupport::Configurable
       extend ActiveSupport::Concern
 
+      # Configuration options may also be set in
+      # Serializers and Adapters
```
Comment (bf4) [inline lib/active_model/serializer/configuration.rb]: No longer correct if we keep api_objects.rb around

File: lib/active_model/serializer/adapter/json_api.rb [hunk 26]
```diff
@@ -21,11 +22,16 @@ def initialize(serializer, options = {})
...
-            serializable_hash_for_collection(options)
-          else
-            serializable_hash_for_single_resource(options)
-          end
+          hash =
+            if serializer.respond_to?(:each)
+              serializable_hash_for_collection(options)
+            else
+              serializable_hash_for_single_resource(options)
+            end
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: someday we'll reify these two latent objects..

**Comment by beauby** [comment]
@bf4 Let's pair on this one when you have time and get it over with.

**Comment by NullVoxPopuli** [comment]
:+1:

File: lib/active_model/serializer/adapter/json_api/api_objects.rb [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+            end
+
+            def include_object?
+              ActiveModel::Serializer.config.jsonapi_include_toplevel_object
+            end
+
+            def object
+              object = {
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: change to `@object ||= `

File: lib/active_model/serializer/adapter/json_api/api_objects.rb (hunk 24 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: return hash

File: lib/active_model/serializer/adapter/json_api/api_objects.rb (hunk 23 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/api_objects.rb]: move back in for now

File: lib/active_model/serializer/adapter/json_api.rb [hunk 28]
```diff
@@ -21,11 +22,16 @@ def initialize(serializer, options = {})
...
+
+          ApiObjects::JsonApi.add!(hash)
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: do we prefer this after?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 29]
```diff
@@ -5,6 +5,7 @@ class JsonApi < Base
         extend ActiveSupport::Autoload
         autoload :PaginationLinks
         autoload :FragmentCache
+        autoload :ApiObjects
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]:
```ruby
def config
  ActiveModel::Serializer.config.jsonapi
end
```

then `ActiveModel::Serializer::Adapter::JsonApi.version = '1.0'`

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: then update usage

File: docs/general/configuration_options.md [hunk 30]
```diff
@@ -9,3 +9,11 @@ The following configuration options can be set on `ActiveModel::Serializer.confi
 ## JSON API
 
 - `jsonapi_resource_type`: Whether the `type` attributes of resources should be singular or plural. Possible values: `:singular, :plural`. Default: `:plural`.
+- `jsonapi_include_toplevel_object`: Whether to include a [top level JSON API member](http://jsonapi.org/format/#document-jsonapi-object)
+   in the response document.
+   Default: `false`.
+- Used when `jsonapi_include_toplevel_object` is `true`:
+  - `jsonapi_version`: The latest version of the spec the API conforms to.
+    Default: `'1.0'`.
+  - `jsonapi_toplevel_meta`: Optional metadata. Not included if empty.
```
Comment (bf4) [inline docs/general/configuration_options.md]: update config to encapsulate in json_api

**Comment by bf4** [comment]
@beauby squashed to one commit, set you as the 'committer' `git log --pretty='%cn' -n1 HEAD` shows you (I'm the 'author')

didn't do the jsonapi.config as it was funky and not high value

File: lib/active_model/serializer/adapter/json_api.rb [hunk 31]
```diff
@@ -6,6 +6,40 @@ class JsonApi < Base
...
         autoload :FragmentCache
 
+        # TODO: if we like this abstraction and other API objects to it,
+        # then extract to its own file and require it.
+        module ApiObjects
+          module JsonApi
+            ActiveModel::Serializer.config.jsonapi_version = '1.0'
+            ActiveModel::Serializer.config.jsonapi_toplevel_meta = {}
+            # Make JSON API top-level jsonapi member opt-in
+            # ref: http://jsonapi.org/format/#document-top-level
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: @beauby  :)

File: lib/active_model/serializer/adapter/json_api.rb [hunk 32]
```diff
@@ -6,6 +6,40 @@ class JsonApi < Base
...
+            def add!(hash)
+              hash.merge!(object) if include_object?
+            end
+
+            def include_object?
+              ActiveModel::Serializer.config.jsonapi_include_toplevel_object
+            end
+
+            # TODO: see if we can cache this
+            def object
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: How about something like:
```ruby
def object
  return @object if @object
  object = {
    jsonapi: {
      version: ActiveModel::Serializer.config.jsonapi_version,
      meta: ActiveModel::Serializer.config.jsonapi_toplevel_meta
    }
  }
  object[:jsonapi].reject! { |_, v| v.blank? }

  @object = object
  @object
end
```

File: docs/general/configuration_options.md (hunk 30 above)
Comment (bf4) [inline docs/general/configuration_options.md]: rename to `jsonapi_toplevel_object_meta`?

Comment (bf4) [inline docs/general/configuration_options.md]: top level meta wouldn't be set in a global config.. but name could be confusing

Comment (beauby) [inline docs/general/configuration_options.md]: Agreed.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 32 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: since we add meta in some of the tests, I'd have to write a way to clear the cache, which I don't want to do right now, could be in a test helper I guess.

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: which is why I left it as a todo :)
